packages/manager/src/managers/versions/VersionsManager.ts

```diff
@@ -0,0 +1,121 @@
+import * as fs from "node:fs/promises";
+import * as path from "node:path";
+import semver from "semver";
+
+import {
+	fetchGitHubReleaseBodyForRelease,
+	GitHubReleaseMetadata,
+} from "../../lib/fetchGitHubReleaseBodyForRelease";
+import { fetchNPMPackageVersions } from "../../lib/fetchNPMPackageVersions";
+
+import { SLICE_MACHINE_GITHUB_PACKAGE_NAME } from "../../constants/SLICE_MACHINE_GITHUB_PACKAGE_NAME";
+import { SLICE_MACHINE_GITHUB_REPOSITORY_NAME } from "../../constants/SLICE_MACHINE_GITHUB_REPOSITORY_NAME";
+import { SLICE_MACHINE_GITHUB_REPOSITORY_ORGANIZATION } from "../../constants/SLICE_MACHINE_GITHUB_REPOSITORY_ORGANIZATION";
+import { SLICE_MACHINE_NPM_PACKAGE_NAME } from "../../constants/SLICE_MACHINE_NPM_PACKAGE_NAME";
+import { VERSION_KIND } from "../../constants/VERSION_KIND";
+
+import { BaseManager } from "../BaseManager";
+
+import { Version } from "./types";
+
+const detectVersionBumpKind = (
+	to: string,
+	from?: string,
+): typeof VERSION_KIND[keyof typeof VERSION_KIND] => {
+	if (!from) {
```

Comment:
> what if from and to are equivalent ?

Good catch 👍

I added a case to handle equivalent versions:

```ts
if (semver.eq(to, from)) {
	return undefined;
} else if (semver.satisfies(to, `~${from}`)) {
	return VERSION_KIND.PATCH;
} else if (semver.satisfies(to, `^${from}`)) {
	return VERSION_KIND.MINOR;
} else {
	return VERSION_KIND.MAJOR;
}
```

In the UI, no badge is displayed if the version is not a bump.

> also does this handle pre-release versions?

It should handle pre-release versions since it uses the official `semver` package. We don't display pre-release versions in the UI and they are filtered out from the versions list before we run `detectVersionBumpKind()`.